doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+      also be trying to make use of the same socket.
+
+    - There are some circumstances where an application might want to multiplex
+      other protocols onto the same UDP socket, for example with protocols like
+      RTP/RTCP or STUN; this can be facilitated using the QUIC fixed bit.
+      However, these use cases cannot be supported without explicit assistance
+      from a QUIC implementation and this use case cannot be facilitated by
+      simply sharing a network socket, as incoming datagrams will not be routed
+      correctly. (We may offer some functionality in future to allow this to be
+      coordinated but this is not for MVP.) Thus this also is not a concern.
+      Moreover, it is extremely unlikely that any such applications are using
+      sockets in blocking mode anyway.
+
+Advantages:
+
+  - An application retains full control of its event loop in non-blocking mode.
+
+    When using libssl in application-level blocking mode, via a custom poller
+    interface, the application would actually able to exercise more control over
+    I/O than it actually is at present when using libssl in blocking mode.
+
+  - Feasible to implement and already working in tests.
+    Minimises further development needed to ship.
+
+  - Does not rely on creating threads and can support blocking I/O at the
+    application level without relying on thread assisted mode.
+
+  - Does not require an application-provided network-side custom BIO to be
+    reworked to support concurrent calls to it.
+
+  - Allows performance-optimal implementation of QUIC RFC requirements.
+
+  - Ensures our internal I/O architecture remains flexible for future evolution
+    without breaking compatibility in the future.
+
+Use of Internal Non-Blocking I/O
+--------------------------------
+
+Based on the above evaluation, implementation has been undertaken using
+non-blocking I/O internally. Applications can use blocking or non-blocking I/O
+at the libssl API level. Network-level BIOs must operate in a non-blocking mode
+or be configurable by QUIC to this end.
+
+![Block Diagram](images/quic-io-arch-1.png "Block Diagram")
+
+### Support of arbitrary BIOs
+
+We need to support not just socket FDs but arbitrary BIOs as the basis for the
+use of QUIC. The use of QUIC with e.g. `BIO_s_dgram_pair`, a bidirectional
+memory buffer with datagram semantics, is to be supported as part of MVP. This
+must be reconciled with the desire to support application-managed event loops.
+
+Broadly, the intention so far has been to enable the use of QUIC with an
+application event loop in application-level non-blocking mode by exposing an
+appropriate OS-level synchronisation primitive to the application. On \*NIX
+platforms, this essentially means we provide the application with:
+
+  - An FD which should be polled for readability, writability, or both; and
+  - A deadline (if any is currently applicable).
```

Comment:
Not sure how this works in practice. How will we create such an FD? Does this work on Windows? Presumably we would need an fd like this for every stream. Does only the connection object need to have deadlines?